### PR TITLE
Change `AsyncOverloadsAvailable` to use `IOperation`

### PR DESCRIPTION
**Pull request by Vannevelj**
Largely there although two unit tests are still failing. Both relate to the surrounding declaration when it is not a regular method (e.g. a global statement or a synchronous lambda)

### CODE REVIEW

**Comment by Vannevelj** [comment]
@Youssef1313 I'd be interested in hearing your thoughts on these two workarounds I came up with. Is there a better way to achieve what I need?

* [`dab3873` (#240)](https://github.com/Vannevelj/SharpSource/pull/240/commits/dab3873b52dedacaf651e0f31eaf065d7faa3fff)
* [`836c866` (#240)](https://github.com/Vannevelj/SharpSource/pull/240/commits/836c866a16a22ed934cdaced5bd7173aa09fb132)

File: SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs [hunk 1]
```diff
@@ -27,84 +25,69 @@ public override void Initialize(AnalysisContext context)
...
-                CheckIfOverloadAvailable(identifierName, context, surroundingDeclaration, invocation);
-                break;
-            case GenericNameSyntax genericName:
-                CheckIfOverloadAvailable(genericName, context, surroundingDeclaration, invocation);
-                break;
-            default:
-                break;
-        }
+            var cancellationTokenSymbol = compilationContext.Compilation.GetTypeByMetadataName("System.Threading.CancellationToken");
+            if (cancellationTokenSymbol is not null)
```
Comment (Youssef1313) [inline SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs:32]: Does this symbol have to exist for the analyzer to report a diagnostic? If not, I think you can just proceed even with a null symbol

File: SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs [hunk 2]
```diff
@@ -27,84 +25,69 @@ public override void Initialize(AnalysisContext context)
...
+        // If the surrounding method is a global statement is it considered as `IsAsync: false` even though async calls work
+        // Using CanBeReferencedByName to try and distinguish between the two scenarios since the top-level statements entry point can't be referenced 
+        if (surroundingMethod is { IsAsync: false, CanBeReferencedByName: true })
         {
             return;
         }
 
-        var surroundingMethodDeclaration = context.SemanticModel.GetDeclaredSymbol(surroundingDeclaration) as IMethodSymbol;
-        foreach (var overload in relevantOverloads)
+        var invocations = context.OperationBlocks.SelectMany(block => block.Descendants().OfType<IInvocationOperation>());
```
Comment (Youssef1313) [inline SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs:53]: I think you can:

```csharp
context.RegisterOperationAction(context =>
{
    var invocation = (IInvocationOperation)context.Operation;
    ...
}, OperationKind.Invocation);
```

File: SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs [hunk 3]
```diff
@@ -27,84 +25,69 @@ public override void Initialize(AnalysisContext context)
...
+            if (isInsideSyncLambda)
+            {
+                continue;
+            }
+
+            var invokedMethodName = invocation.TargetMethod.Name;
+            var invokedTypeName = invocation.TargetMethod.ContainingType.Name;
+
+            var methodsInInvokedType = invocation.TargetMethod.ContainingType.GetMembers().OfType<IMethodSymbol>();
+            var relevantOverloads = methodsInInvokedType.Where(x => x.Name == $"{invokedMethodName}Async");
```
Comment (Youssef1313) [inline SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs:67]:
```suggestion
            var relevantOverloads = invocation.TargetMethod.ContainingType.GetMembers($"{invokedMethodName}Async").OfType<IMethodSymbol>();
```

File: SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs [hunk 4]
```diff
@@ -43,8 +43,9 @@ private static void Analyze(OperationBlockAnalysisContext context, INamedTypeSym
             return;
         }
 
-        // Problem: If the surrounding method is a global statement is it considered as `IsAsync: false` even though async calls work
-        if (surroundingMethod is { IsAsync: false })
+        // If the surrounding method is a global statement is it considered as `IsAsync: false` even though async calls work
+        // Using CanBeReferencedByName to try and distinguish between the two scenarios since the top-level statements entry point can't be referenced 
+        if (surroundingMethod is { IsAsync: false, CanBeReferencedByName: true })
```
Comment (Youssef1313) [inline SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs:48]: If you want to be very specific about the top-level statement case, you can compare the name against `WellKnownMemberNames.TopLevelStatementsEntryPointMethodName`

Comment (Vannevelj) [reply SharpSource/SharpSource/Diagnostics/AsyncOverloadsAvailableAnalyzer.cs:48]: Excellent, that's what I was hoping for

**Pull request merged**